Collapse range deletions

ajkr opened the pull request:
Added a tombstone-collapsing mode to RangeDelAggregator, which eliminates overlap in the TombstoneMap. In this mode, we can check whether a tombstone covers a user key using upper_bound() (i.e., binary search). However, the tradeoff is the overhead to add tombstones is now higher, so at first I've only enabled it for range scans (compaction/flush/user iterators), where we expect a high number of calls to ShouldDelete() for the same tombstones. Point queries like Get() will still use the linear scan approach.

Also in this diff I changed RangeDelAggregator's TombstoneMap to use multimap with user keys instead of map with internal keys. Callers sometimes provided ParsedInternalKey directly, from which it would've required string copying to derive an internal key Slice with which we could search the map.

Test Plan: unit tests

ajkr commented:
ping

yiwu-arbug commented on db/range_del_aggregator.cc:
```diff
@@ -115,6 +133,159 @@ Status RangeDelAggregator::AddTombstones(
+              RangeTombstone(
+                  Slice(), Slice(),
+                  std::max(
+                      untermed_seq == kMaxSequenceNumber ? 0 : untermed_seq,
```
set `untermed_seq` to 0 whenever you set it to `kMaxSequenceNumber` to save a check here?

ajkr replied:
oh, untermed_seq == 0 means we just covered an existing point with seqnum 0, in which case we need line 240 to evaluate to true so we can force insertion of the new point. but I think this logic will be obsolete now with your idea to split the existing ranges when they overlap a new point.

yiwu-arbug commented on db/range_del_aggregator.cc:
```diff
@@ -115,6 +133,159 @@ Status RangeDelAggregator::AddTombstones(
+            *new_range_dels_iter_end, *tombstone_map_iter_end);
+      }
+
+      if (new_to_old_start_cmp < 0) {
```
I feel the logic here can be simplify. What if break it into three steps:
1. break the existing range that cover new_range_start into two.
2. break the existing range that cover new_range_end into two.
3. update all ranges that are fully cover by the new range.

ajkr replied:
great idea, thanks :)

yiwu-arbug replied:
If we're going to keep a vector of new range instead of one range at a time, then the current logic looks better, though I still feel it hard to understand an probably can be simplify.

ajkr commented on db/range_del_aggregator.cc:
```diff
@@ -115,6 +133,159 @@ Status RangeDelAggregator::AddTombstones(
+    // until the next tombstone starts. For gaps between real tombstones and
+    // for the last real tombstone, we denote end keys by inserting fake
+    // tombstones with sequence number zero.
+    std::vector<RangeTombstone> new_range_dels{
```
one of the motivations for using a vector of new points is that, in a future optimization, we're considering merging multiple new points simultaneously (e.g., all points from a snapshot stripe in an SST file that we just read), which should be faster than merging new points one-at-a-time.

ajkr commented on db/range_del_aggregator.cc:
```diff
@@ -50,6 +58,14 @@ bool RangeDelAggregator::ShouldDelete(const ParsedInternalKey& parsed) {
   }
   const auto& tombstone_map = GetTombstoneMap(parsed.sequence);
+  if (collapse_deletions_) {
+    auto iter = tombstone_map.upper_bound(parsed.user_key.ToString());
```
1. yes, not sure what I was thinking
2. lower_bound gives us the first point >= search key, but we want the last point <= search key.

ajkr commented on db/range_del_aggregator.cc:
```diff
@@ -115,6 +133,159 @@ Status RangeDelAggregator::AddTombstones(
+    // raising the seqnum for the to-be-inserted element (we insert the max
+    // seqnum between the next new interval and the unterminated interval).
+    SequenceNumber untermed_seq = kMaxSequenceNumber;
+    SequenceNumber prev_seen_seq = 0;
```
sure

ajkr commented:
Sorry I missed your general questions earlier:

1. RangeDelAggregator is constructed for every read operation, whether it's a get or an iterator. For iteration there is one RangeDelAggregator managed at the top-level, which is either DBIter (iterators for users) or CompactionIterator (iterators for flush or compaction). The lower-level iterators add tombstones into it, e.g., each time an SST file is read (https://github.com/facebook/rocksdb/blob/master/db/table_cache.cc#L243). Also the logic for reading values is updated to check whether keys are covered by range deletions with ShouldDelete().

2. The order is kind of arbitrary. For example, the Get() code path uses FilePicker to decide which file to check next for a key. We add the range deletions for that file just before checking it for the key.

yiwu-arbug commented on db/range_del_aggregator.cc:
```diff
@@ -179,7 +179,6 @@ Status RangeDelAggregator::AddTombstone(RangeTombstone tombstone) {
     // raising the seqnum for the to-be-inserted element (we insert the max
     // seqnum between the next new interval and the unterminated interval).
     SequenceNumber untermed_seq = kMaxSequenceNumber;
-    SequenceNumber prev_seen_seq = 0;
```
good job getting rid of this variable. it make the state transition simpler.

ajkr commented:
Addressing your comments:

1. I think for the complexity analysis we also have to account for the scanning of tombstones whose key-ranges are fully contained in the new tombstone's key-range. This is for modifying/erasing tombstones that are now covered. Maybe it's unfair to claim this is an O(n^2) scan for each tombstone since tombstones will probably be similar sizes, but I think in the worst case it will be.
2,3. Sure, I like your idea of using set with custom comparator :).
4. Oh, but we actually use tombstones with seq 0 (note there can be multiple, not just at the end, in case there are gaps between tombstones) as sentinels to denote end key. For example, if we want to store one tombstone with begin_key="a", end_key="b", and seq=10, then the map will contain {"a"->RangeTombstone(seq=10), "b"->RangeTombstone(seq=0)}. I thought this might be simpler because we don't have to maintain end_key since each tombstone extends until the next one starts.

yiwu-arbug commented:
1. A range full covered by the new tombstone will be remove only once. Well, a range can be split but the ranges we end up with be O(N). (N = number of total input ranges). So the amortize cost for a range is O(logN) (logN for removal from a balanced binary search tree).

But well, I know the actual running time needs to be benchmark with a workload.. And O(logN) operation on std::map takes a big constant factor.

ajkr commented:
1. True it can only be removed once, but it may be considered by many scans if it has a higher seqnum than the new tombstones being added---i.e., it's not guaranteed to be removed just because its key-range is fully covered. Anyways, I can leave this for now and we can revisit later for further optimization, ok?

yiwu-arbug commented:
I see. Yeah I forget about the seq_id > new seq_id case. Sure, go ahead. All the above are to share my two cents and have a little bit discussion.

ajkr commented:
Thanks.

2,3. I tried this, but unfortunately ran into the issue that set's keys are supposed to be immutable. We could mark end_key_ and seq_ as "mutable" in the RangeTombstone struct, but I don't think it'd be intuitive to readers why start_key_ is immutable and others aren't. So, I think it's cleaner to continue with map for now.

yiwu-arbug commented:
cool.
